### PR TITLE
Deprecate uptimes in pong message

### PR DESCRIPTION
**Pull request by ceyonur**
# Why this should be merged

Closes: #1361 

# How this works

Adds SubnetUptimes field to Ping message, handle old/new ping messages according to existence of SubnetUptimes field (exist = new, otherwise = old). This is guaranteed to exist for new messages (even for non-validators) because we forcefully append primary network uptimes to that list.

Adds deprecation comments to Pong message type/functions.

# How this was tested

Locally tested with old + new versions.

### CODE REVIEW

File: network/peer/peer.go [hunk 1]
```diff
@@ -702,7 +706,84 @@ func (p *peer) handle(msg message.InboundMessage) {
...
+				zap.Uint32("uptime", uptime),
+			)
+			p.StartClose()
+			return
+		}
+		p.observeUptime(subnetID, uptime)
+	}
+}
+
+// // Deprecated, TODO remove this in the future.
```
Comment (danlaine) [inline network/peer/peer.go:785]:
```suggestion
// Deprecated, TODO remove this in the future.
```

**Review by danlaine — APPROVED** [review]

File: network/peer/peer.go [hunk 2]
```diff
@@ -702,7 +706,84 @@ func (p *peer) handle(msg message.InboundMessage) {
...
+			)
+			// Make sure we always include the primary network ID
+			// in the ping message.
+			if subnetID != constants.PrimaryNetworkID {
+				continue
+			}
+			subnetUptime = 0
+		}
+
+		copyID := subnetID
```
Comment (joshua-kim) [inline network/peer/peer.go:739]: Why do we need to copy this?

File: network/peer/peer.go [hunk 3]
```diff
@@ -702,7 +706,84 @@ func (p *peer) handle(msg message.InboundMessage) {
...
+				zap.Uint32("uptime", uptime),
+			)
+			p.StartClose()
+			return
+		}
+		p.observeUptime(subnetID, uptime)
+	}
+}
+
+// // Deprecated: remove this in the future.
```
Comment (joshua-kim) [inline network/peer/peer.go:785]:
```suggestion
// Deprecated: remove this in the future.
```

File: network/peer/peer.go (hunk 2 above)
Comment (ceyonur) [reply network/peer/peer.go:739]: otherwise it escapes at `subnetID[:]`

File: proto/p2p/p2p.proto [hunk 4]
```diff
@@ -64,9 +64,14 @@ message Message {
 // Message that the local node sends to its remote peers,
 // in order to periodically check its uptime.
 //
+// Deprecated: remove this comment in the future.
 // On receiving "ping", the remote peer responds with the observed
 // uptime value of the message sender in "pong" message.
-message Ping {}
+message Ping {
+  // subnet_uptimes contains subnet uptime percentages, including the primary
+  // network.
```
Comment (StephenButtolph) [inline proto/p2p/p2p.proto:72]: Can we separate out the primary network into its own field (like in pong) to save ~33 bytes per ping message

Comment (ceyonur) [reply proto/p2p/p2p.proto]: done

File: network/peer/peer.go [hunk 5]
```diff
@@ -702,7 +706,114 @@ func (p *peer) handle(msg message.InboundMessage) {
...
+		if err != nil {
+			p.Log.Debug("failed to get peer uptime percentage",
+				zap.Stringer("nodeID", p.id),
+				zap.Stringer("subnetID", subnetID),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		copyID := subnetID
```
Comment (StephenButtolph) [inline network/peer/peer.go:738]: We typically write this as:
```suggestion
		subnetID := subnetID
```
in other places

File: network/peer/peer.go [hunk 6]
```diff
@@ -702,7 +706,114 @@ func (p *peer) handle(msg message.InboundMessage) {
...
+	// it is backwards compatible
+	// TODO: remove this in the future
+	if msg.GetUptime() != 0 {
+		p.handleNewPing(msg)
+	} else {
+		p.handleOldPing(msg)
+	}
+}
+
+func (p *peer) createPingMessage() (message.OutboundMessage, error) {
```
Comment (StephenButtolph) [inline network/peer/peer.go:723]: Can we share this code with `handleOldPing`? They are doing almost exactly the same thing

File: network/peer/peer.go (hunk 5 above)
Comment (ceyonur) [reply network/peer/peer.go]: done

File: network/peer/peer.go (hunk 6 above)
Comment (ceyonur) [reply network/peer/peer.go]: done

File: message/ops.go [hunk 7]
```diff
@@ -20,7 +20,7 @@ type Op byte
 const (
 	// Handshake:
 	PingOp Op = iota
-	PongOp
+	PongOp    // Deprecated: remove this in the future.
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply message/ops.go]: yup changed it. but we still need a way to indicate we're sending the deprecated pong message so peers won't overwrite any of their primary uptimes with 0. see:  
https://github.com/ava-labs/avalanchego/pull/1362/files#diff-6a2c510edabd05f1f93bfd1964b7bb3822e0d4b808e1651b3183f3bb285039f2R816-R821

[the remark this answers is not in the document]
Comment (StephenButtolph) [reply message/ops.go]: I updated the PR to align the ping and pong message handled more closely (doesn't need the flag)

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**